Registry Entry Attachment API

Leo40Git opened the pull request:
This PR adds the _Registry Entry Attachment_ API module, which allows associating arbitrary values with any object, provided said object is managed by a registry.

- [x] Code-based path
  - [x] Getting attachment values
  - [x] Setting attachment values
- [x] Data-based path 
  - [x] Syncing to client (depends on #34)
- [x] Javadoc!

Leo40Git committed:
Registry Item Attributes API

Half-completed code-based path, no data-based path yet

Leo40Git committed:
Get data-based path working

Leo40Git committed:
Make resource reloader async

Values are now deserialized during the load phase rather than the apply phase to allow for better error messages

Leo40Git committed:
Implement code-based path for setting attributes

Attributes set via code are "built-in", and are always overriden by values set via data

Leo40Git committed:
Added methods for common attribute types

BasiqueEvangelist commented on library/data/registry-item-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/RegistryItemAttributeReloader.java:
```diff
@@ -0,0 +1,170 @@
+		return CompletableFuture.supplyAsync(() -> {
+			Map<RegistryItemAttribute<?, ?>, AttributeMap> attributeMaps = new HashMap<>();
+			for (Map.Entry<? extends RegistryKey<? extends Registry<?>>, ? extends Registry<?>> entry : Registry.REGISTRIES.getEntries()) {
+				Identifier registryId = entry.getKey().getValue();
+				StringBuilder pathSB = new StringBuilder();
+				if (!"minecraft".equals(registryId.getNamespace()))
+					pathSB.append(registryId.getNamespace()).append('/');
+				pathSB.append(registryId.getPath());
+				profiler.push(ID + "/finding_resources/" + pathSB);
+				Collection<Identifier> jsonIds = manager.findResources("attributes/" + pathSB, s -> s.endsWith(".json"));
+				if (jsonIds.isEmpty())
+					continue;
+				profiler.pop();
+				BuiltinRegistryItemAttributeHolder<?> holder = RegistryItemAttributeHolderImpl.getBuiltin(entry.getValue());
+				for (Identifier jsonId : jsonIds) {
+					Identifier attribId = getAttributeId(jsonId);
+					RegistryItemAttribute<?, ?> attrib = holder.getAttribute(attribId);
+					if (attrib == null) {
+						LOGGER.warn("Unknown attribute {} (from {})", attribId, jsonId);
+						continue;
+					}
+					profiler.push(ID + "/getting_resources{" + jsonId + "}");
+					List<Resource> resources;
+					try {
+						resources = manager.getAllResources(jsonId);
+					} catch (IOException e) {
+						LOGGER.error("Failed to get all resources for {}", jsonId);
+						LOGGER.catching(Level.ERROR, e);
+						continue;
+					}
+					AttributeMap attribMap = attributeMaps.computeIfAbsent(attrib,
+							id -> new AttributeMap(entry.getValue(), attrib));
+					profiler.swap(ID + "/processing_resources{" + jsonId + "," + attribId + "}");
+					for (Resource resource : resources) {
+						attribMap.processResource(resource);
+					}
+					profiler.pop();
+				}
+			}
+			return new LoadedData(attributeMaps);
```
This long block of text makes the method unreadable. Could you add some empty lines to separate logical blocks of code?

LambdAurora replied:
In the old codestyle there was mandatory blank lines between different indentation levels.

Leo40Git replied:
I added comments. Does that work?

BasiqueEvangelist replied:
That's a bit better, but imo you should still add some blank lines

BasiqueEvangelist replied:
Nice! It's much more readable now.

Leo40Git committed:
Add comments so the code is less... uh, dense

Leo40Git committed:
Spaces and braces

Leo40Git committed:
god I suck at organizing code

Leo40Git committed:
Initial package-info

BasiqueEvangelist commented on library/data/registry-item-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/package-info.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * A simple API for adding arbitrary values to any object instances managed by a {@link net.minecraft.util.registry.Registry Registry}.<p>
+ *
+ * {@link org.quiltmc.qsl.registry.attribute.api.RegistryItemAttribute RegistryItemAttributes} contains a number of methods for creating
```
Why does this rename `RegistryItemAttribute` to `RegistryItemAttributes`?

Leo40Git replied:
That's an error, oops.

Leo40Git committed:
Add dispatched utilities

Misc style/JD fixes

OroArmor commented:
Maybe Item - > Entry since item has a more common definition

BasiqueEvangelist commented on library/data/registry-item-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/QuiltRegistryInternals.java:
```diff
@@ -0,0 +1,9 @@
+	BuiltinRegistryItemAttributeHolder<?> qsl$getBuiltinItemAttributeHolder();
+	void qsl$setBuiltinItemAttributeHolder(BuiltinRegistryItemAttributeHolder<?> holder);
+
+	BuiltinRegistryItemAttributeHolder<?> qsl$getItemAttributeHolder();
+	void qsl$setItemAttributeHolder(BuiltinRegistryItemAttributeHolder<?> holder);
```
This should be called `qsl$getDataItemAttributeHolder`, since it's unclear that it's supposed to hold the data attributes and not the combined attribute holder

Leo40Git committed:
Item -> Entry, part 1

Leo40Git committed:
Item -> Entry, part 2

Leo40Git committed:
Item -> Entry, part 3

Leo40Git committed:
Rename QuiltRegistryInternals methods to be less confusing

Leo40Git committed:
Stick combined attribute holder in RegistryMixin

i509VCB commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/BuiltinRegistryEntryAttributeHolder.java:
```diff
@@ -0,0 +1,22 @@
+	private final HashMap<Identifier, RegistryEntryAttribute<R, ?>> attributes;
+
+	public BuiltinRegistryEntryAttributeHolder() {
+		super();
```
Super call is implicit always

i509VCB commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/RegistryEntryAttributeReloader.java:
```diff
@@ -0,0 +1,210 @@
+	private String getAttributeMapPath(Identifier registryId) {
+		String path = registryId.getPath();
+		if (!"minecraft".equals(registryId.getNamespace()))
+			path = registryId.getNamespace() + "/" + path;
```
Sir this is java

Leo40Git replied:
I have no idea what this means

Leo40Git replied:
Well whatever this method died so I guess this is resolved?

i509VCB commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/package-info.java:
```diff
@@ -0,0 +1,34 @@
+ *
+ * <h3>Data Driven Capabilities</h3>
+ *
+ * All attributes declared using this API can be set using a data pack! Simply create the following file:<br>
```
Include an example of one of these data driven files.

i509VCB commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/mixin/RegistryMixin.java:
```diff
@@ -0,0 +1,53 @@
+import org.quiltmc.qsl.registry.attribute.impl.*;
+import net.minecraft.util.registry.Registry;
+
+@Mixin(Registry.class)
```
Should this be getting applied to Registry or MutableRegistry?

Leo40Git replied:
Attributes can be attached to any element without a restriction on how it was registered, so Registry seems correct.

i509VCB reviewed (commented):
My review isn't done yet, I need to think about if this is part of the right module and if there is a better way to do this on the scope of registries

Leo40Git committed:
i5 review

package-info: Add example for attribute map file
BuiltinRegistryEntryAttributeHolder: Remove explicit super call

Leo40Git committed:
L I C E N S E

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryEntryAttribute.java:
```diff
@@ -0,0 +1,92 @@
+	static <R, T extends DispatchedType> RegistryEntryAttribute<R, T> createDispatched(RegistryKey<Registry<R>> registryKey, Identifier id,
+																					   Map<String, Codec<? extends T>> codecs) {
+		return createDispatched(registryKey, id, codecs, null);
+	}
```
These methods should probably use `Identifier` as the map key instead of `String`

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryExtensions.java:
```diff
@@ -0,0 +1,34 @@
+												  Consumer<AttributeSetter<R>> setterConsumer) {
+		Registry.register(registry, id, item);
+		setterConsumer.accept(new BuiltinRegistryEntryAttributeSetterImpl<>(registry, item));
+	}
```
This should return `item` back to behave similarly to the original `Registry.register` method.

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/package-info.java:
```diff
@@ -0,0 +1,42 @@
+ * {@link org.quiltmc.qsl.registry.attribute.api.RegistryEntryAttribute RegistryEntryAttribute} contains a number of methods for creating
+ * a new attribute.<p>
+ *
+ * To retrieve an attribute value, fist get the attribute holder using
```
```suggestion
 * To retrieve an attribute value, first get the attribute holder using
```

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/mixin/RegistryMixin.java:
```diff
@@ -0,0 +1,69 @@
+			qsl$combinedAttributeHolder = new CombinedRegistryEntryAttributeHolder<>(
+					(RegistryEntryAttributeHolder<Object>) qsl$dataAttributeHolder,
+					(RegistryEntryAttributeHolder<Object>) qsl$builtinAttributeHolder);
+		}
```
Won't `CombinedRegistryEntryAttributeHolder` still work even if both the delegate and fallback are null?

Leo40Git replied:
Good point! Yes it would! I'm a fool!

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/RegistryEntryAttributeHolderImpl.java:
```diff
@@ -0,0 +1,84 @@
+			return Optional.ofNullable(attribute.getDefaultValue());
+		} else {
+			return Optional.of((T) rawValue);
+		}
```
Can't you just use `Table.get` here?

Leo40Git replied:
for some reason I thought `Table` didn't have a method like that??

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/package-info.java:
```diff
@@ -0,0 +1,42 @@
+ *
+ * The {@code values} object functions as a map, mapping registry entry IDs to attribute values.<p>
+ *
+ * Example file, for an {@code Integer} attribute for the {@code Item} registry:<pre><code>
```
Can you also add the path of this example file?

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/RegistryEntryAttributeReloader.java:
```diff
@@ -0,0 +1,226 @@
+		return CompletableFuture.runAsync(() -> data.apply(profiler), executor);
+	}
+
+	// "attributes/<registry_path>" for Minecraft registries,
```
Why does this drop the `minecraft` in Minecraft registries' paths? IMO this is just confusing.

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/RegistryEntryAttributeReloader.java:
```diff
@@ -0,0 +1,226 @@
+			AttributeMap attribMap = attributeMaps.computeIfAbsent(attrib,
+					id -> new AttributeMap(registry, attrib));
+			for (var resource : resources) {
+				attribMap.processResource(resource);
```
This should probably catch all `IllegalArgumentException`s and `JsonParseExceptions`

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/RegistryEntryAttributeReloader.java:
```diff
@@ -0,0 +1,226 @@
+		public void processResource(Resource resource) {
+			JsonObject obj = JsonHelper.deserialize(new InputStreamReader(resource.getInputStream()));
+			boolean replace = JsonHelper.getBoolean(obj, "replace", false);
+			JsonObject values = JsonHelper.getObject(obj, "values");
```
`JsonHelper.getObject` throws if the property isn't present

Leo40Git committed:
Update library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/package-info.java

Co-authored-by: BasiqueEvangelist <basiqueevangelist@yandex.ru>

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/BuiltinRegistryEntryAttributeHolder.java:
```diff
@@ -0,0 +1,37 @@
+		attributes = new HashMap<>();
+	}
+
+	public <T> void registerAttribute(RegistryEntryAttribute<R, T> attribute) {
+		attributes.put(attribute.getId(), attribute);
+	}
+
+	public RegistryEntryAttribute<R, ?> getAttribute(Identifier id) {
+		return attributes.get(id);
+	}
```
Why is the `Identifier -> RegistryEntryAttribute` map stored in the `BuiltinRegistryEntryAttributeHolder`? It would make sense if it actually used the map, but since it's only used by the resource reloader, this should probably be moved somewhere else

Leo40Git replied:
Moved to `QuiltRegistryInternals`/`RegistryMixin`.

Sidenote: `BuiltinRegistryEntryAttributeHolder` was supposed to be used for _only_ the built-in holder... but the data-driven holder still uses it because I forgot to actually check.

Leo40Git committed:
Basique review

DispatchedType: type is now an Identifier
RegistryExtensions.registerWithAttributes: returns registered entry
Removed EmptyRegistryEntryAttributeHolder, since Combined works if both are null
Replace dumb Table.rowMap().get() with Table.get()
package-info: Extended data-driven example
RegistryEntryAttributeReloader: namespace is no longer dropped from registry IDs
RegistryEntryAttributeReloader: catch exceptions from JSON parsing
Yeet BuiltinRegistryEntryAttributeHolder, instead registered attributes are stored directly in QuiltRegistryInternals/RegistryMixin

Leo40Git committed:
RegistryEntryAttribute: add methods for String attributes

Leo40Git committed:
CombinedRegistryEntryAttributeHolder: respect default value when both are null

Leo40Git committed:
RegistryEntryAttribute.createDispatched: Use Function instead of Map

Using Map implies that it gets stored - which it doesn't (AFAIK, anyway)

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/RegistryEntryAttributeHolderImpl.java:
```diff
@@ -0,0 +1,91 @@
+	public <T> Optional<T> getValue(R item, RegistryEntryAttribute<R, T> attribute) {
+		var rawValue = valueTable.get(item, attribute);
+		if (rawValue == null) {
+			return Optional.ofNullable(attribute.getDefaultValue());
```
This should now return `Optional.empty()`, since this will hide the built-in attribute value if there is a default value and the default value'll be handled by the `CombinedRegistryEntryAttributeHolder` anyway
```suggestion
			return Optional.empty();
```

Leo40Git replied:
Oop, good catch!

Leo40Git committed:
CombinedRegistryEntryAttributeHolder: Fix built-in value always being hidden by default value

Leo40Git committed:
javadoc!

Leo40Git commented:
I officially mark this PR as ready to review!

Leo40Git committed:
forgot some C's

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/RegistryEntryAttributeReloader.java:
```diff
@@ -0,0 +1,229 @@
+					return;
+				}
+
+				if (values == null) {
+					LOGGER.error("Missing 'values' element in {}, ignoring file", resource.getId());
+					return;
+				}
```
How would `values` be null here? It's guaranteed by `JsonHelper.getObject` to be not null (since that'll throw if there is no `"values"` key)

Leo40Git committed:
remove useless null check

TheGlitch76 commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/package-info.java:
```diff
@@ -0,0 +1,43 @@
+ * to get the value assigned to the specified registry entry.<p>
+ *
+ * <p>
+ *
```
Formatting police here. You have two paragraph breaks in a row.
```suggestion
 *
```

TheGlitch76 commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/package-info.java:
```diff
@@ -0,0 +1,43 @@
+ *
+ * <ul>
+ *     <li>Attribute values set in-code (using
+ *     {@linkplain org.quiltmc.qsl.registry.attribute.api.RegistryExtensions#registerWithAttributes(net.minecraft.util.registry.Registry, net.minecraft.util.Identifier, java.lang.Object, java.util.function.Consumer) this extension method})
```
"this extension method" is not helpful to me reading the documentation, because I have to click a link to figure out what you're talking about. Just use a normal link here.

Leo40Git committed:
package-info fixes

Leo40Git committed:
RegistryEntryAttributeHolderImpl: Optimize table storage

Instead of mapping entry -> attribute -> value, it now maps attribute -> entry -> value
This makes it so that we have a lesser amount of large maps, rather than a greater amount of small maps

Leo40Git committed:
I love Java generics

Haven-King commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryEntryAttribute.java:
```diff
@@ -0,0 +1,275 @@
+	 * @return default value
+	 */
+	@Nullable T getDefaultValue();
+}
```
A method `Optional<T> get(R entry)` would make sense to me here as a much cleaner alternative to `RegistryEntryAttributeHolder.get(Registry.ITEM_KEY).getValue(ATTRIBUTE, ENTRY)`.

Leo40Git replied:
Added as a `default` method.

Haven-King commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryEntryAttribute.java:
```diff
@@ -0,0 +1,275 @@
+ * @param <R> registry entry type
+ * @param <T> value type
+ */
+public interface RegistryEntryAttribute<R, T> {
```
All of the `create` methods here should have overrides that take a `Registry<R>` instead of a `RegistryKey<Registry<R>>`. Between the two, I'd much rather have the `Registry<R>` methods than the `RegistryKey<Registry<R>>` ones.

Leo40Git replied:
I should just make these things store the `Registry` instance directly honestly, registry keys are apparently really annoying to work with (generics, go figure).

Leo40Git replied:
So, guess what? I did exactly that. If this is a bad idea, tell me.

Haven-King commented on library/data/registry-entry-attributes/src/testmod/java/org/quiltmc/qsl/registry/attributes/test/SimpleAttributeTest.java:
```diff
@@ -0,0 +1,70 @@
+import net.minecraft.util.registry.Registry;
+import net.minecraft.world.World;
+
+public class SimpleAttributeTest implements ModInitializer {
```
I'd like to see a test for the dispatched attributes as well, both here in code and in JSON.

Leo40Git replied:
Done!

Haven-King reviewed (changes requested):
Looks like a good start.

Leo40Git committed:
Add RegistryEntryAttribute.getValue utility method

Also rename value type parameter to V because it makes sense in my ape brain

Leo40Git committed:
All my homies hate RegistryKey

Attributes now store the registry they're attached to directly rather than its key
The attribute won't work if the registry isn't present, so using keys offers no advantage AFAIK

Also more of that "RegistryEntryAttribute<V>" refactor from last commit oops

Leo40Git committed:
Forgot to refactor SimpleAttributeTest, oops

Leo40Git committed:
Fix this embarrassing control flow error

Leo40Git committed:
Add dispatched/polymorphic attribute test

Leo40Git committed:
Forgot to upadte this method's JD

Leo40Git committed:
fix indent

Leo40Git committed:
Add command to dump builtin values

Leo40Git committed:
RegistryEntryAttributeReloader: registry item -> registry entry

Leo40Git committed:
license apply thing

Leo40Git committed:
DispatchedType: fix JD link to createDispatched

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/package-info.java:
```diff
@@ -0,0 +1,41 @@
+ * A simple API for adding arbitrary values to any object instances managed by a {@link net.minecraft.util.registry.Registry Registry}.<p>
+ *
+ * {@link org.quiltmc.qsl.registry.attribute.api.RegistryEntryAttribute RegistryEntryAttribute} contains a number of methods for creating
+ * a new attribute.<p>
```
"for creating a new attribute" doesn't sound right. Maybe this instead?
```suggestion
 * {@link org.quiltmc.qsl.registry.attribute.api.RegistryEntryAttribute RegistryEntryAttribute} contains a number of methods for creating
 * new attributes.<p>
```

i509VCB commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryEntryAttribute.java:
```diff
@@ -0,0 +1,289 @@
+ * @param <R> registry entry type
+ * @param <V> value type
+ */
+public interface RegistryEntryAttribute<R, V> {
```
Is this intended for public extension? Given the minimality of the implementation, it may make more sense to make this a final class?

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/DumpBuiltinAttributesCommand.java:
```diff
@@ -0,0 +1,192 @@
+		var registry = RegistryArgumentType.getRegistry(ctx, "registry");
+		try {
+			execute0(ctx, registry);
+		} catch (RuntimeException e) {
+			LOGGER.error("Uncaught exception occurred", e);
+			throw UNCAUGHT_EXCEPTION.create();
+		}
```
Won't `CommandManager` catch all exceptions for you anyway?

Leo40Git replied:
AFAIK, it just swallows the exception and doesn't print the stacktrace - catching it manually allows me to do that.

i509VCB commented on library/data/registry-entry-attributes/src/testmod/java/org/quiltmc/qsl/registry/attributes/test/SimpleAttributeTest.java:
```diff
@@ -0,0 +1,69 @@
+		@Override
+		public TypedActionResult<ItemStack> use(World world, PlayerEntity user, Hand hand) {
+			if (!world.isClient) {
+				int one = TEST_ATTRIBUTE.getValue(this)
+						.orElseThrow(() -> new RuntimeException(TEST_ATTRIBUTE + " not set via datapack!"));
+				float two = TEST_ATTRIBUTE_2.getValue(this)
+						.orElseThrow(() -> new RuntimeException(TEST_ATTRIBUTE_2 + " not set via built-in!"));
```
Given the behavior I see with TEST_ATTRIBUTE_2, the exception being thrown where `TEST_ATTRIBUTE` is not present is an impossible condition? Reason I thought that way was the 2nd attribute checks if the built-in default value was set, but the setter also supplies a default value for the 1st attribute.

Leo40Git replied:
oh right oops

Testing this properly would require exposing the data-driven/built-in split in code, which seems like a needless implementation detail to me (or I could just remove the built-in default value).

i509VCB commented on library/data/registry-entry-attributes/src/testmod/java/org/quiltmc/qsl/registry/attributes/test/DispatchedAttributeTest.java:
```diff
@@ -0,0 +1,137 @@
+import net.minecraft.world.World;
+import java.util.Map;
+
+public class DispatchedAttributeTest implements ModInitializer {
```
Hmm this is definitely an interesting use case. I feel the docs in api would be much nicer given this example and explaining how dispatched types would allow for functionally composable items, blocks, block entities, etc.

i509VCB commented on library/data/registry-entry-attributes/src/testmod/java/org/quiltmc/qsl/registry/attributes/test/DispatchedAttributeTest.java:
```diff
@@ -0,0 +1,137 @@
+import net.minecraft.world.World;
+import java.util.Map;
+
+public class DispatchedAttributeTest implements ModInitializer {
```
I like final things

BasiqueEvangelist replied:
What?

i509VCB commented on library/data/registry-entry-attributes/src/testmod/java/org/quiltmc/qsl/registry/attributes/test/DispatchedAttributeTest.java:
```diff
@@ -0,0 +1,137 @@
+	public static abstract class FuncValue implements DispatchedType {
+		// in a real-world application, you'd probably use a Registry for this
+		public static final Map<Identifier, Codec<? extends FuncValue>> CODECS = Util.make(() ->
+				ImmutableMap.<Identifier, Codec<? extends FuncValue>>builder()
```
Ah only if type erasure wasn't so mean lol

mschae23 replied:
What do you mean? Vanilla always uses registries for dispatched codecs.

i509VCB replied:
Immutable map builders always cause javac to get confused

i509VCB commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/DispatchedType.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * This allows for polymorphic attribute types!
+ */
+public interface DispatchedType {
```
I feel the purpose of this would be better explained using part of the testmod you have made for this.

Leo40Git replied:
I've tried to improve this interface's Javadoc using the testmod as an example.

i509VCB commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryEntryAttribute.java:
```diff
@@ -0,0 +1,289 @@
+/**
+ * Represents an attribute that is attached to a registry entry. Maps to a value in a {@link RegistryEntryAttributeHolder}.
+ *
+ * @param <R> registry entry type
+ * @param <V> value type
```
These are a little confusing?

`R` is the type of the object in the registry and `V` is the attached data type?

Leo40Git replied:
Cleared this javadoc up a little.

i509VCB commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryEntryAttribute.java:
```diff
@@ -0,0 +1,289 @@
+	 * @param entry registry entry
+	 * @return attribute value, or empty if no value is assigned
+	 */
+	default Optional<V> getValue(R entry) {
```
I personally dislike Optionals, but my guess is there is a reason Optionals were chosen here?

Leo40Git replied:
null bad

I don't remember honestly, I could just make this return `V`

Leo40Git replied:
OH RIGHT, I made it `Optional` so primitive types are a bit easier to deal with (IE you don't need to interface with boxed types, just use an `orElseX` method)

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/RegistryEntryAttributeReloader.java:
```diff
@@ -0,0 +1,225 @@
+		@SuppressWarnings("unchecked")
+		private <R, V> void applyOne(RegistryEntryAttribute<R, V> attrib, AttributeMap attribMap) {
+			var registry = attrib.getRegistry();
+			assert registry != null : "bruh momento";
```
Maybe use `Objects.requireNonNull` instead of an assert?
~~also is this assert even necessary?~~
```suggestion
			Objects.requireNonNull(registry, "registry");
```

i509VCB replied:
Assert is something you would use in a dev environment to ensure some value makes sense. For failing at runtime requireNonNull is more appropriate.

i509VCB commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryExtensions.java:
```diff
@@ -0,0 +1,63 @@
+	 * @param <R> registry entry type
+	 */
+	@FunctionalInterface
+	public interface AttributeSetter<R> {
```
I think this would probably be better described as a `DefaultAttributesBuilder`.

i509VCB commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryEntryAttribute.java:
```diff
@@ -0,0 +1,289 @@
+	 */
+	static <R> RegistryEntryAttribute<R, Long> createLong(Registry<R> registry, Identifier id) {
+		return create(registry, id, Codec.LONG, null);
+	}
+
+	/**
+	 * Creates a new {@code float} property.
+	 *
+	 * @param registry registry this attribute is tied to
+	 * @param id identifier
+	 * @param <R> registry entry type
+	 * @return a new attribute
+	 */
+	static <R> RegistryEntryAttribute<R, Float> createFloat(Registry<R> registry, Identifier id) {
+		return create(registry, id, Codec.FLOAT, null);
+	}
+
+	/**
+	 * Creates a new {@code double} property.
+	 *
+	 * @param registry registry this attribute is tied to
+	 * @param id identifier
+	 * @param <R> registry entry type
+	 * @return a new attribute
+	 */
+	static <R> RegistryEntryAttribute<R, Double> createDouble(Registry<R> registry, Identifier id) {
+		return create(registry, id, Codec.DOUBLE, null);
+	}
+
+	/**
+	 * Creates a new {@code String} property.
+	 *
+	 * @param registry registry this attribute is tied to
+	 * @param id identifier
+	 * @param <R> registry entry type
+	 * @return a new attribute
+	 */
+	static <R> RegistryEntryAttribute<R, String> createString(Registry<R> registry, Identifier id) {
+		return create(registry, id, Codec.STRING, null);
+	}
```
I feel it would be more clear if any of these functions with default values or functional getters were named differently. This would make it more clear to a user of the API why the extra arguments exist.

Leo40Git replied:
The functional getters are exclusive to the "dispatched" utility methods, though I could rename all the default value variants to `createXWithDefault`

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/RegistryEntryAttributeHolderImpl.java:
```diff
@@ -0,0 +1,96 @@
+	}
+
+	public static <R> RegistryEntryAttributeHolder<R> getCombined(Registry<R> registry) {
+		return getInternals(registry).qsl$getCombinedAttributeHolder();
```
This will return `null` if no mod set an attribute for this registry.

i509VCB commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryEntryAttribute.java:
```diff
@@ -0,0 +1,289 @@
+	}
+
+	/**
+	 * Creates a new {@code boolean} property.
```
```suggestion
	 * Creates a new {@code boolean} property with no default value.
```

We should probably note the implicit contract this function declares and all others like it.

i509VCB commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/package-info.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * <h2>Registry Item Attributes</h2>
```
I want to think about the name of the module's primary use case for a bit.

Attribute feels a bit weird since it describes a characteristic of an object. Sure this is the case with dispatched attributes where those could have functional characteristics. However with things like booleans, it isn't clearly a characteristic of an object but rather a piece of attached data.

Leo40Git replied:
Hm. Maybe "Registry Data Attachment API"?
- `RegistryEntryAttribute` -> `RegistryDataKey`
- `RegistryEntryAttributeHolder` -> `RegistryDataMap`

i509VCB reviewed (dismissed):
I quite like the idea behind dispatched attributes and how those open up the door to very nicely composable items, blocks, block entities, etc.

Some polish will probably be needed before I can say this is done and ready to merge.

One question I've had lingering in my mind: is this data intended to be seen by the logical client or only the logical server?

Leo40Git committed:
Apply suggestions from code review

Co-authored-by: i509VCB <i509vcb@gmail.com>
Co-authored-by: BasiqueEvangelist <basiqueevangelist@yandex.ru>

Leo40Git committed:
Merge RegistryEntryAttribute and its implementation

Leo40Git committed:
Clear up default vs no default attribute factory methods

Leo40Git committed:
RegistryExtensions: Rename AttributeSetter to DefaultAttributesBuilder

Leo40Git committed:
Fix null attribute holder if no attributes are registered for a registry

Leo40Git committed:
Make type var javadoc a bit clearer

Leo40Git committed:
SimpleAttributeTest: Fix impossible failure condition

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryEntryAttribute.java:
```diff
@@ -0,0 +1,322 @@
+	 */
+	public static <R, V> RegistryEntryAttribute<R, V> createWithDefault(Registry<R> registry, Identifier id, Codec<V> codec,
+																		@Nullable V defaultValue) {
+		return new RegistryEntryAttribute<>(registry, id, codec, defaultValue);
```
This doesn't actually register the attribute to the registry

Leo40Git replied:
i should really stop doing big refactors late at night

Leo40Git committed:
Fix attributes not being registered on creation

Leo40Git committed:
Fix missing import here

Leo40Git committed:
Forgot to update this, oops

Leo40Git committed:
Improve DispatchedType JD

Leo40Git committed:
Register our custom registry argument type

I have no idea if this is required or not since the game did run before I added this, but there seems to be validation for this, so...

Leo40Git committed:
right, this should be a codec registry

Leo40Git committed:
Add NotExtendable to AttributeHolder

Leo40Git committed:
Hide RegistryEntryAttributeHolder as impl detail

Leo40Git committed:
Nullable annotations

Leo40Git committed:
RegistryExtensions.registerWithAttributes: fix JD

Leo40Git commented:
I have just been informed that Minecraft does not just sync datapack files, and in fact has specific sync packets for everything data-driven.

I'll add packets for syncing the stuff once we have a Networking API, whether that happens before or after this PR is merged.

Leo40Git committed:
Client-side attributes

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryEntryAttribute.java:
```diff
@@ -0,0 +1,233 @@
+import java.util.function.Function;
+
+/**
+ * Represents an attribute that is attached to a registry entry. Maps to a value in a {@link RegistryEntryAttributeHolder}.
```
`RegistryEntryAttributeHolder` is an internal class, so it shouldn't be mentioned here

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/CombinedRegistryEntryAttributeHolder.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.Optional;
+
+@SuppressWarnings("ClassCanBeRecord")
+public final class CombinedRegistryEntryAttributeHolder<R> implements RegistryEntryAttributeHolder<R> {
```
Is `CombinedRegistryEntryAttributeHolder` even needed anymore? Couldn't you just put its logic in the `RegistryEntryAttribute.getValue` method?

Leo40Git replied:
combined has been yeeted

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/package-info.java:
```diff
@@ -0,0 +1,39 @@
+ * <h3>Data-Driven Capabilities</h3>
+ *
+ * All attributes declared using this API can be set using a data pack! Simply create the following file:<br>
+ * {@code data/<attribute_namespace>/attributes/<registry_key_path>/<attribute_name>.json}<p>
```
This should now also say how you can specify attributes with resource packs

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/DumpBuiltinAttributesCommand.java:
```diff
@@ -0,0 +1,196 @@
+import static net.minecraft.server.command.CommandManager.argument;
+import static net.minecraft.server.command.CommandManager.literal;
+
+public final class DumpBuiltinAttributesCommand {
```
This command should also probably put attributes in different directories depending on side

Leo40Git committed:
forgot to register attributes _again_

also removed link to impl interface

Leo40Git committed:
Fix testmod package

Leo40Git committed:
Fix common attribute values possibly being loaded from assets

Leo40Git committed:
Fix assets attributes always being cleared

Add testmod
Yeet CombinedRegistryEntryAttributeHolder in favor of putting the behavior into RegistryEntryAttribute.getValue
Merge RegistryEntryAttributeHolder and its implementation

Leo40Git committed:
Add missing javadoc

Leo40Git committed:
DumpBuiltinAttributesCommand: dump to assets/data depending on side

Leo40Git committed:
Update package-info.java

BasiqueEvangelist commented on library/data/registry-entry-attributes/src/testmod/java/org/quiltmc/qsl/registry/attribute/test/mixin/client/DebugHudMixin.java:
```diff
@@ -0,0 +1,28 @@
+	@SuppressWarnings({"rawtypes", "unchecked"})
+	@Inject(method = "getRightText", at = @At(value = "INVOKE", target = "Ljava/util/List;add(Ljava/lang/Object;)Z", ordinal = 2,
+			shift = At.Shift.AFTER), locals = LocalCapture.CAPTURE_FAILHARD)
+	public void qsl$addTestAttribute(CallbackInfoReturnable<List<String>> cir, long l, long m, long n, long o, List list,
+									 BlockPos blockPos, BlockState blockState) {
+		var dList = (List<String>) list;
```
Can't you just say that the `list` argument is a `List<String>`?

Leo40Git replied:
...oh yeah, forgot generics don't exist at runtime. I thought Mixin would shout at me for trying that.

Leo40Git committed:
Client testmod: remove dumb cast

Rename attribute

Leo40Git committed:
DispatchedType: fix example snippet

Leo40Git committed:
Use ResourceType instead of ambiguous boolean

Leo40Git committed:
Add missing license headers

Leo40Git committed:
a

Leo40Git committed:
Mark impl classes as internal

OroArmor commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryEntryAttribute.java:
```diff
@@ -0,0 +1,336 @@
+ * @param <R> type of the entries in the registry
+ * @param <V> attached value type
+ */
+@SuppressWarnings("ClassCanBeRecord")
```
Why should this class not be a record?

Leo40Git replied:
IIRC, this class was originally supposed to be used in an identity-based map, so you couldn't just create 2 attributes with the same key and have them be "merged". Now that a standard map is used, it can be a record (well, the implementation class can be, anyways).

OroArmor commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryEntryAttribute.java:
```diff
@@ -0,0 +1,336 @@
+			return Optional.of(value);
+		}
+		return Optional.ofNullable(defaultValue);
+	}
```
I feel like this is more implementation than actual api. I would change it so that this class is an interface, while theimplementation of this new interface are in the impl package.

OroArmor commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/DumpBuiltinAttributesCommand.java:
```diff
@@ -0,0 +1,199 @@
+
+@ApiStatus.Internal
+public final class DumpBuiltinAttributesCommand {
+	public static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
```
I feel like this command class shouldnt be exposed in game or be allowed to be registered

OroArmor commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/mixin/CommandManagerMixin.java:
```diff
@@ -0,0 +1,42 @@
+
+	@Inject(method = "<init>", at = @At("RETURN"))
+	public void qsl$registerDumpCommand(CommandManager.RegistrationEnvironment environment, CallbackInfo ci) {
+		DumpBuiltinAttributesCommand.register(dispatcher);
```
Yeah from above. I would make sure this command can only be called if it is a dev environment

Leo40Git replied:
I don't know, this command seems useful to datapack developers, and I doubt they'd want to go spin up a dev environment just to use it.

Maybe lock it behind a system property, so it can be enabled via `-D`?

OroArmor replied:
that sounds like a good compromise

OroArmor commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/mixin/RegistryMixin.java:
```diff
@@ -0,0 +1,76 @@
+
+@Mixin(Registry.class)
+public abstract class RegistryMixin<R> implements QuiltRegistryInternals<R> {
+	@Unique private final HashMap<Identifier, RegistryEntryAttribute<R, ?>> attributes = new HashMap<>();
```
`Map` not `HashMap` for the type

OroArmor commented on library/data/registry-entry-attributes/src/testmod/java/org/quiltmc/qsl/registry/attribute/test/DispatchedAttributeTest.java:
```diff
@@ -0,0 +1,137 @@
+		}
+	}
+
+	// built-in of one type
```
this can be a javadoc. not super useful, but it looks nicer

OroArmor commented on library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryExtensions.java:
```diff
@@ -0,0 +1,63 @@
+ * Extensions for working with {@link Registry}s.
+ */
+public final class RegistryExtensions {
+	/**
+	 * Registers an entry into a registry, also setting built-in attribute values for it.
+	 *
+	 * @param registry target registry
+	 * @param id entry identifier
+	 * @param entry entry
+	 * @param builderConsumer builder consumer
+	 * @param <R> type of the entries in the registry
+	 * @param <T> type of the entry we're currently registering (may be a subclass of {@code R})
+	 * @return the newly registered entry
+	 */
+	public static <R, T extends R> T registerWithAttributes(Registry<R> registry, Identifier id, T entry,
+												  Consumer<DefaultAttributesBuilder<R>> builderConsumer) {
+		Registry.register(registry, id, entry);
+		builderConsumer.accept(new DefaultAttributesBuilderImpl<>(registry, entry));
+		return entry;
+	}
```
Not sure how useful this method is. I can see its use case in mods, but it kinda seems like the only way to add attributes to entries with default values.

Leo40Git replied:
That's because it is, mostly to let the registering mod determine what its entry should and should not have.

Ideally, if you were using an attribute, you'd have code to handle an entry which doesn't have a value assigned to said attribute, whether that would be a default value or an extra calculation that you'd cache yourself.

I _could_ add a `setAttributes` method here that allows you to mess with the attributes of an already-registered entry, if people want something like that.

OroArmor reviewed (changes requested):
Seems pretty good. Just some small nitpicks

Leo40Git committed:
Split impl details from RegistryEntryAttribute

Leo40Git committed:
Prevent attributes from being replaced

Add method to retrieve already-registered attribute

Leo40Git committed:
Weaken typing here

Leo40Git committed:
Convert test comments to javadoc

Leo40Git committed:
Fix compilation

Leo40Git committed:
Stick Internal annotation on attribute impl class

Leo40Git committed:
Snakify folder name

Leo40Git committed:
Snakify module name as well

(forgot it should be synced with folder name)

Leo40Git committed:
Add missing value function

Remove default value getter from interface, it's an impl detail now

Leo40Git committed:
Snakify this dependency

Leo40Git committed:
Lock "dump builtin attributes" command behind system property

Add "-Dquilt.data.registry_entry_attributes.dumpbuiltin_command=true" to your run args to re-enable it

Leo40Git committed:
Correct mixin prefix

Leo40Git committed:
Don't use default value if missing value function is set

Leo40Git committed:
RegistryEntryAttributeReloader: missing continue

Leo40Git committed:
Use Boolean.getBoolean

Leo40Git committed:
Fix up licenses

Leo40Git commented:
I officially declare this PR as 1.18 compatible.

Leo40Git committed:
Fix licenses AGAIN

Leo40Git committed:
excuse me mr. backslash how tf did you get there

Leo40Git committed:
murder spaces in licenses

Leo40Git committed:
Make raw Javadoc easier on the eyes

LambdAurora commented on library/data/registry_entry_attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/package-info.java:
```diff
@@ -0,0 +1,46 @@
+ * {
+ *   "replace": false,
+ *   "values": {
+ *     "minecraft:netherite_pickaxe": 20
```
I've been told on Discord it's only `"key": value`, but that omit optional values, which is quite important.

Leo40Git replied:
I have implemented optional values! I think.

Leo40Git committed:
Add support for optional entries

Leo40Git committed:
Remove Quilt Loader dependency

LambdAurora commented on library/data/registry_entry_attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/RegistryEntryAttributeImpl.java:
```diff
@@ -0,0 +1,87 @@
+											   @Nullable Function<R, V> missingValueFunction)
+		implements RegistryEntryAttribute<R, V> {
+	@Override
+	public Optional<V> getValue(R entry) {
```
Please have empty lines before and after a new scope, here everything being so close together makes it hard to read.

LambdAurora commented on library/data/registry_entry_attributes/src/main/resources/quilt_registry_entry_attributes.mixins.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "required": true,
+  "package": "org.quiltmc.qsl.registry.attribute.mixin",
+  "compatibilityLevel": "JAVA_16",
```
```suggestion
  "compatibilityLevel": "JAVA_17",
```

LambdAurora commented on library/data/registry_entry_attributes/src/testmod/resources/quilt_registry_entry_attributes_testmod.mixins.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "required": true,
+  "package": "org.quiltmc.qsl.registry.attribute.test.mixin",
+  "compatibilityLevel": "JAVA_16",
```
```suggestion
  "compatibilityLevel": "JAVA_17",
```

LambdAurora commented on library/data/registry_entry_attributes/src/main/java/org/quiltmc/qsl/registry/attribute/mixin/CommandManagerMixin.java:
```diff
@@ -0,0 +1,42 @@
+import net.minecraft.server.command.CommandManager;
+import net.minecraft.server.command.ServerCommandSource;
+
+// FIXME replace with command registration event once that's implemented
```
This might have issues if command module isn't in data or core library.

Leo40Git replied:
...do we seriously not have any writeups in place for an optional dependency on another library's module?

LemmaEOF replied:
I'm not sure anyone's considered optional deps yet.

Leo40Git committed:
Update Mixin compatability level

Co-authored-by: LambdAurora <aurora42lambda@gmail.com>

Leo40Git committed:
add a newline

Leo40Git committed:
Prevent use of assets-based attributes on dedicated server

(technically they exist fine on it, they just never get a value since there's no assets reloader)

Leo40Git committed:
Fix this check to not break everything oops

Leo40Git commented on library/data/registry_entry_attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/RegistryEntryAttributeReloader.java:
```diff
@@ -61,6 +61,9 @@ public static void register(ResourceType source) {
 	private final ResourceType source;
 
 	private RegistryEntryAttributeReloader(ResourceType source) {
+		if (source == ResourceType.SERVER_DATA) {
```
(I already fixed this, oops)

Leo40Git committed:
limcensses

Leo40Git committed:
Make reloader ID more descriptive

Make data and assets reloaders have different IDs

Leo40Git committed:
Start working on tag support

Leo40Git committed:
Pretty up AttributeMap a bit

License stuff

Leo40Git committed:
Fix "dump builtin attrs" command

No longer always fails on dedicated servers if a client attribute is registered

Leo40Git committed:
AttributeTarget.Tagged: abstract no more!

Leo40Git committed:
checkstyle

Leo40Git committed:
Add missing metadata to fabric.mod.json

Leo40Git committed:
Fix up javadoc

Leo40Git committed:
AttributeMap: make handleArray method less stupid

Rather than using an enhanced for loop AND keeping track of an index, just use a standard for loop you dingus

LemmaEOF commented on library/data/registry_entry_attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/reloader/RegistryEntryAttributeReloader.java:
```diff
@@ -0,0 +1,208 @@
+	}
+
+	@Override
+	public <T> Tag<T> getTag(RegistryKey<? extends Registry<T>> registryKey, Identifier id) {
```
To check, is this to expose REAs as tags, or to use tags as keys to REAs? My prior art, [MCDict](https://github.com/CottonMC/MCDict), let you use tags as keys but I never really got around to exposing dicts as tags. That'd require mixing into tag loading to check REA keys as well.

Leo40Git replied:
This is for using tags as keys to REAs. Currently unimplemented because tags are scary.

LemmaEOF replied:
MCDict lazy-loads tags to work around some issues with that: https://github.com/CottonMC/MCDict/commit/6916b311d1ececfbf0a505c5b3b9c05d37083373

LemmaEOF commented on library/data/registry_entry_attributes/build.gradle:
```diff
@@ -0,0 +1,15 @@
+			"resource_loader"
+	])
+	interLibraryDependencies([
+			"tags"
```
Tags are part of the data lib, aren't they? wouldn't that mean it's an intra-library dependency?

LambdAurora replied:
And REA is part of the data lib too, which means it's fine.

LemmaEOF replied:
Yeah, I'm just confused with it using `inter` instead of `intra` here I guess. It's saying the opposite of what it means. That's probably a problem for a different PR though.

LambdAurora replied:
Oh true, did not notice, weird.

LemmaEOF commented on library/data/registry_entry_attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/DispatchedType.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Utility interface used for {@link RegistryEntryAttribute#dispatchedBuilder(Registry, Identifier, Function)}.
+ *
+ * <p>This allows for polymorphic attribute types!<br>
```
👀👀👀

LemmaEOF commented on library/data/registry_entry_attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryEntryAttribute.java:
```diff
@@ -0,0 +1,308 @@
+ * @param <R> type of the entries in the registry
+ * @param <V> attached value type
+ */
+public interface RegistryEntryAttribute<R, V> {
```
I keep misreading `Entry` in this as `Entity`, which might mix it up with EAMs if other folks have the same issue. Maybe a simpler name would be a good idea? I just called them Dicts back in [my implementation of this concept](https://github.com/CottonMC/MCDict) ages ago.

LemmaEOF commented on library/data/registry_entry_attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryEntryAttribute.java:
```diff
@@ -0,0 +1,308 @@
+	/**
+	 * Specifies on what side this attribute should exist.
+	 */
+	enum Side {
```
Should this have the same options as or use [`TagType`](https://github.com/QuiltMC/quilt-standard-libraries/blob/1.18/library/data/tags/src/main/java/org/quiltmc/qsl/tag/api/TagType.java)?

Leo40Git replied:
Maybe? I'll look into that later.

LemmaEOF commented on library/data/registry_entry_attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryExtensions.java:
```diff
@@ -0,0 +1,65 @@
+	 *
+	 * @param registry        target registry
+	 * @param id              entry identifier
+	 * @param entry           entry
```
Might want a better description on this - maybe call it `toRegister` or something? It's not super clear that This Is What You Are Registering from the name/docs

LemmaEOF reviewed (approved):
I'm super excited for this, tbh. It's gonna be *super* helpful to have this built into QSL instead of being a third-party lib like mine was.

Depending on how hands-on we want to be, it might be a good idea to convert furnace fuel times and composting chances into REAs in QSL instead of having a third-party mod do it. We can take a look at it another time maybe.

Overall, this is looking pretty good! I've left a few notes that might want to be addressed before merge.

Leo40Git committed:
Possibly implement tag support

This is completely untested!

Leo40Git committed:
Rename to Registry Dictionaries API

Leo40Git committed:
Redo "missingValueFunction"

Leo40Git commented:
This API has been renamed to _Registry Dictionary API_!

Leo40Git committed:
Propagate rename to Javadoc and logging, too

Actually rename the folder dict maps are loaded from (it was changed in the javadoc, but not in the code itself)
also rename a bunch of local vars

Leo40Git committed:
Add test for tags

Leo40Git committed:
Fix assets-based dicts not loading

Turns out you can't depend on tags loading for resource stuff...
New issue: can't reference data tags from assets since assets get loaded first

Fix SimpleDictTest

Leo40Git committed:
FORGOT A LICENSE AAA

Leo40Git committed:
Fix optional entries and tags

Leo40Git committed:
Mark RegistryDict as non-extendable

Leo40Git committed:
Make defaultValue and computeFunction explicitly mutually exclusive

Leo40Git committed:
IDEA lied to me :(

Leo40Git committed:
RegistryDict: add value class

For now, this is only used by RegistryDict.get

Leo40Git committed:
if only applyLicenses wasn't broken

Leo40Git committed:
fix buildscript

Leo40Git committed:
DumpBuiltinDictsCommand: use Command module

remove custom RegistryArgumentType and use vanilla IdentifierArgumentType instead

Leo40Git committed:
Sync v1

Leo40Git committed:
limcensme

Leo40Git commented:
Server-to-client synchronization of dictionaries has been added! This PR is now pretty much complete.

Leo40Git committed:
specify config for compileOnly dep

Haven-King commented on library/data/registry_dicts/src/main/java/org/quiltmc/qsl/registry/dict/api/RegistryExtensions.java:
```diff
@@ -0,0 +1,65 @@
+	 * @param <T>             type of the entry we're currently registering (may be a subclass of {@code R})
+	 * @return the newly registered entry
+	 */
+	public static <R, T extends R> T registerWithDictValues(Registry<R> registry, Identifier id, T toRegister,
+															Consumer<DefaultDictBuilder<R>> builderConsumer) {
+		Registry.register(registry, id, toRegister);
+		builderConsumer.accept(new DefaultDictBuilderImpl<>(registry, toRegister));
+		return toRegister;
+	}
```
The builder pattern here is strange and unintuitive. I'd rather have a large number of methods with various numbers of parameters, similar to the JDK's `Map#of` methods:
```java
public <T> register(Registry<R> registry, Identifier id, T toRegister, RegistryDict<R, V> dict, V value) {
  ...
}

public <T> register(Registry<R> registry, Identifier id, T toRegister, RegistryDict<R, V> dict1, V value1, RegistryDict<R, V> dict2, V value2) {
  ...
}

...
```
In a world where transparent transformations with CHASM exists, providing util classes as part of a public facing API would be strange. When CHASM is eventually able to be used in production, these kinds of static helper methods should probably just be made part of the `Registry` class itself.

Leo40Git replied:
```java
public <R, T extends R, V1> T register(Registry<R>, Identifier, T, RegistryDict<R, V1>, V1);
public <R, T extends R, V1, V2> T register(Registry<R>, Identifier, T, RegistryDict<R, V1>, V2, RegistryDict<R, V2>, V2);
```
Not really scalable, unless I want to go for unchecked varargs... I do agree my Kotlin builder pattern shenanigans are ugly, though.

Haven-King replied:
It's not terribly scalable, no, but I don't expect many items to be registered with more than 3-4 default attachments. Providing up to 7-8 methods similar to those shown above should be sufficient. If a mod *really* needs a larger number of default attachments they can write their own helper methods :smile:

Leo40Git replied:
Another idea would be to scrap the "built-in values _must_ be set at registration time" limitation completely and just expose a `setBuiltin(RegistryDict<R>, R, V)` helper method...

Haven-King replied:
I think that built-in values being set at registration time does make sense. With that in mind, maybe just provide the `of` methods described above *in addition to* the existing builder pattern.

Leo40Git replied:
I've added overloads up to 4 dict/value pairs. Is that enough?

Leo40Git replied:
I've added overloads up to 4 dict/value pairs. Is that enough?

Haven-King replied:
Probably

Haven-King commented on library/data/registry_dicts/src/testmod/java/org/quiltmc/qsl/registry/dict/test/DispatchedDictTest.java:
```diff
@@ -0,0 +1,152 @@
+import org.quiltmc.qsl.registry.dict.api.RegistryDict;
+import org.quiltmc.qsl.registry.dict.api.RegistryExtensions;
+
+public class DispatchedDictTest implements ModInitializer {
```
This file is very large. To make it clear and useful as an example, not just a test, I'd break each child class out into its own individual file.

Haven-King commented on library/data/registry_dicts/src/main/java/org/quiltmc/qsl/registry/dict/api/ComputeFunction.java:
```diff
@@ -0,0 +1,50 @@
+
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A function that computes a value for an entry that doesn't have one.
+ *
+ * @param <R> type of registry entry
+ * @param <V> type of value
+ */
```
I'd like the documentation here to say where/why this is used. Even using more descriptive language like `A function that computes a default value for a registry entry that doesn't have one defined explicitly` would probably be enough.

Haven-King commented on library/data/registry_dicts/src/main/java/org/quiltmc/qsl/registry/dict/api/ComputeFunction.java:
```diff
@@ -0,0 +1,50 @@
+	 * @return computed value
+	 * @throws ComputeFailedException if computation fails for some reason.
+	 */
+	@NotNull V computeFor(R entry) throws ComputeFailedException;
+
+	/**
+	 * Thrown by {@link ComputeFunction#computeFor(Object)}, should computation fail.
+	 */
+	class ComputeFailedException extends Exception {
+		public ComputeFailedException(String message) {
+			super(message);
+		}
+
+		public ComputeFailedException(String message, Throwable cause) {
+			super(message, cause);
+		}
+	}
```
Relying on exception handling for logic flow is generally a bad idea. Some sort of `Result<V>` class that can hold either a `V` or an error message would be preferable.

Haven-King commented on library/data/registry_dicts/src/main/java/org/quiltmc/qsl/registry/dict/api/RegistryDict.java:
```diff
@@ -0,0 +1,338 @@
+		 * @return new dictionary
+		 */
+		public RegistryDict<R, V> build() {
+			var dict = new RegistryDictImpl<>(registry, id, valueClass, codec, side, defaultValue, computeFunction);
```
A single implementation having both the default value and compute function capabilities seems strange to me. Not a huge deal, but I might break this into two different implementations (possible with a shared abstract parent class).

Haven-King commented on library/data/registry_dicts/src/main/java/org/quiltmc/qsl/registry/dict/api/ComputeFunction.java:
```diff
@@ -0,0 +1,50 @@
+ * @param <V> type of value
+ */
+@FunctionalInterface
+public interface ComputeFunction<R, V> {
```
A function is by definition something that computes. This class name is therefore meaningless. I might name it something like `DefaultDictionaryValueFunction`. It's a longer name, but that's not necessarily a bad thing.

Haven-King commented on library/data/registry_dicts/src/main/java/org/quiltmc/qsl/registry/dict/api/WrongValueClassException.java:
```diff
@@ -0,0 +1,44 @@
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+
+/**
+ * Thrown by {@link RegistryDict#get(Registry, Identifier, Class)}, should a dictionary with a matching ID be found
+ * but have a different value class than expected.
+ */
```
This class seems unnecessary. In the situations where it would be raised, I'd just throw a runtime error and crash. I can't think of another good, legitimate response to this kind of error.

Haven-King commented on library/data/registry_dicts/src/main/java/org/quiltmc/qsl/registry/dict/impl/RegistryDictHolder.java:
```diff
@@ -0,0 +1,115 @@
+		return getInternals(registry).quilt$getDictEntries();
+	}
+
+	/// impl for RegistryDict.get
+	@SuppressWarnings("unchecked")
+	public static <R, V> @Nullable RegistryDict<R, V> getDict(Registry<R> registry, Identifier id, Class<V> valueClass)
+			throws WrongValueClassException {
+		var dict = getDict(registry, id);
+		if (dict == null) {
+			return null;
+		}
+		if (dict.valueClass() != valueClass) {
+			throw new WrongValueClassException(registry, id, valueClass, dict.valueClass());
+		}
+		return (RegistryDict<R, V>) dict;
+	}
```
This method doesn't seem to ever be called. Another good reason to just delete `WrongValueClassException` :smile:

Leo40Git replied:
It's called by the public method `RegistryDict.get(Registry, Identifier, Class)` - but yeah, I think making it a checked exception was a bit much.

Haven-King commented on library/data/registry_dicts/src/main/java/org/quiltmc/qsl/registry/dict/impl/reloader/DictMap.java:
```diff
@@ -0,0 +1,231 @@
+		return map;
+	}
+
+	// FIXME tag support!
```
Does this need to be fixed?

Leo40Git replied:
Oh, right. I... never actually finished this part. Well, what with tags basically being completely rewritten, might as well do it now!

Leo40Git replied:
...no, wait, I _did_ do tag support! I just forgot to remove this.
But I'll probably end up rewriting tag support because of the 1.18.2 refactor anyway.

Leo40Git replied:
AGH, forgot to remove this again

Haven-King commented on library/data/registry_dicts/src/main/java/org/quiltmc/qsl/registry/dict/impl/reloader/DictMap.java:
```diff
@@ -0,0 +1,231 @@
+import org.quiltmc.qsl.registry.dict.api.RegistryDict;
+
+final class DictMap {
+	private final Registry<?> registry;
+	private final RegistryDict<?, ?> dictionary;
+	private final TagGetter tagGetter;
+	private final Map<DictTarget, Object> map;
```
Is there a good reason this class has zero type safety?

Leo40Git replied:
None that I recall!

Haven-King replied:
Then I'd definitely like to see this get generified prior to merge.

Haven-King commented on library/data/registry_dicts/src/main/java/org/quiltmc/qsl/registry/dict/impl/reloader/DictTarget.java:
```diff
@@ -0,0 +1,109 @@
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+
+sealed interface DictTarget {
```
Why does this need to be sealed and can't just be left package private?

Leo40Git replied:
Okay, yeah, `sealed` is way overkill. Will be removed.

Haven-King reviewed (changes requested):
I think the question that needs to be answered for me most here is why are we describing these as dictionaries? While "dictionary" might describe the structure of the potentially JSON data that's being parsed to create them, these values can be created and added via code as well, in a way that entirely circumvents the "dictionary" aspect of this API. A more accurate name would be something along the lines of `RegistryEntryAttachment` or `RegistryEntryAttribute`. 

This API has definitely received a lot of work and I do think it's close, but there are a number of things that still need to be looked at.

Leo40Git commented:
> A more accurate name would be something along the lines of `RegistryEntryAttachment` or `RegistryEntryAttribute`.

The latter is... the original name for this (well, it was `RegistryItemAttribute` but `Item` got changed out for `Entry` really early on). `Attribute` was dropped since it was too close to _entity_ attributes, but `Attachment` might be a good option?

Haven-King commented:
> The latter is... the original name for this (well, it was `RegistryItemAttribute` but `Item` got changed out for `Entry` really early on). `Attribute` was dropped since it was too close to _entity_ attributes, but `Attachment` might be a good option?

Haha I couldn't remember the original name even though I looked for it. Figures that I would go back to it. Was there a discussion that led to the rename? I do think `RegistryEntryAttachment` is a more descriptive name than `RegistryDictionary` and think it would be worth discussing.

Leo40Git commented:
I have tracked down the start of the discussion for the rename (Greencord message link): https://discord.com/channels/833872081585700874/881659500845752320/917301286263738389

Should probably do a proper thread/poll this time around before I commit to mass renames again...

Leo40Git committed:
ComputeFunction: rename to DefaultValueProvider

use Result wrapper instead of exception for logic flow

Leo40Git committed:
RegistryExtensions: rename registerWithDictValues to register

Add overloads to register to avoid clumsy builder-consumer

Leo40Git committed:
Remove WrongValueClassException

Leo40Git committed:
RegistryDictImpl: split

Leo40Git committed:
RegistryDict: rename compute function in javadoc, too

Leo40Git committed:
DispatchedDictTest: move inner classes to their own files

Leo40Git committed:
Log4J -> SLF4J

Leo40Git committed:
Merge remote-tracking branch 'upstream/1.18' into reg-item-attributes

Leo40Git committed:
Update build.gradle

Use QSL Base's init entrypoints

Leo40Git commented:
FYI, there is currently a _poll_ on the Quilt Development Discord about the new name for this module: https://discord.com/channels/833872081585700874/951957775280373800

Leo40Git committed:
DumpBuiltinDictsCommand: Log4J -> SLF4J

Leo40Git committed:
Fix tag support

DictMap: add type params
RegistryDictReloader, DictMap: Log4J -> SLF4J

Leo40Git committed:
DictMap: remove resolved FIXME

Leo40Git committed:
RegistryDict -> RegistryEntryAttachment

Haven-King commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/api/DefaultValueProvider.java:
```diff
@@ -0,0 +1,100 @@
+		 * {@return <code>true</code> if result is failed, <code>false</code> otherwise}
+		 */
+		public boolean isFailed() {
+			return value != null;
```
This check doesn't seem totally safe. I could feasibly see an attachment having `null` as a valid value (even if it's a little weird). I'd rather have a separate `private final boolean isError` field.

Haven-King commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/api/DefaultValueProvider.java:
```diff
@@ -0,0 +1,100 @@
+		/**
+		 * {@return <code>true</code> if result is failed, <code>false</code> otherwise}
+		 */
+		public boolean isFailed() {
```
`isFailed` isn't a proper English grammar. I'd either make it `hasFailed` or invert it to `isOk` for some more standard language.

Haven-King reviewed (changes requested):
*Almost* there!

Leo40Git committed:
DefaultValueProvider.Result: add explicit hasFailed flag

ComputedDefaultRegistryEntryAttachmentImpl: fix flipped if

Leo40Git committed:
RegistryEntryAttachmentSync: consolidate all sync-related methods

Actually send sync packets (forgot to call an important method, oops)
Add Environment annotations to all client-only classes/methods

LambdAurora commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/ClientInitializer.java:
```diff
@@ -0,0 +1,38 @@
+
+@ApiStatus.Internal
+@Environment(EnvType.CLIENT)
+public final class ClientInitializer implements ClientModInitializer {
```
This should be moved into a client subpackage.

LambdAurora commented on library/data/registry_entry_attachments/src/main/resources/fabric.mod.json:
```diff
@@ -0,0 +1,50 @@
+  ],
+  "depends": {
+    "fabricloader": ">=0.12",
+    "minecraft": ">=1.18",
```
```suggestion
    "minecraft": ">=1.18.2",
```

Leo40Git committed:
Update minimum Minecraft version

Co-authored-by: LambdAurora <aurora42lambda@gmail.com>

Leo40Git committed:
ClientInitializer: move to client subpackage

OroArmor commented on library/data/registry_entry_attachments/build.gradle:
```diff
@@ -0,0 +1,24 @@
+		}
+		command {
+			compileOnly("command")
+			//testmodOnly("command")
```
```suggestion
```

LambdAurora committed:
Merge branch '1.18' into reg-item-attributes

LambdAurora committed:
Convert registry entry attachments to Quilt Loader setup.
